MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,1771 @@
+              }
+              for (size_t k = 0; k < sv0_lonesharedtrack_which_trkidx.size(); k++) {
+                track_vec tks_sv0 = sv_total_track_which_trk_vec[vtxidx0];
+                size_t idx = sv0_lonesharedtrack_which_trkidx[k];
+                h_resolve_shared_jets_lonetrkvtx_dphi->Fill(fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)));
+
+                // drop the lone track pointing backwards from the vertex direction!
+                if (fabs(reco::deltaPhi(tks_sv0[idx]->phi(), phi0)) > M_PI / 2) {
+                  eraseElement(sv_total_track_which_trkidx[vtxidx0], idx);
+                }
+              }
+              if (verbose) {
+                printf("sv0'idx: %lu with ntrack before: %u",vtxidx0, sv0.nTracks());
+                printf("sv1'idx: %lu with ntrack before: %u",vtxidx1, sv1.nTracks());
+              }
+
+              track_set  sv0_resolved_sharedtracks_trkset;
+              for (unsigned int i = 0; i < sv_total_track_which_trkidx[vtxidx0].size(); ++i) {
+                int idx = sv_total_track_which_trkidx[vtxidx0][i];
+                track_vec tks_sv0 = sv_total_track_which_trk_vec[vtxidx0];
+                sv0_resolved_sharedtracks_trkset.insert(tks_sv0[idx]);
+
+              }
+              std::vector<reco::TransientTrack> sv0_resolved_sharedtracks_ttks;
+              for (auto tk : sv0_resolved_sharedtracks_trkset)
+                sv0_resolved_sharedtracks_ttks.push_back(tt_builder->build(tk));
+
+              reco::Vertex sv0_resolved_sharedtracks;
+
+              for (const TransientVertex& tv : kv_reco_dropin(sv0_resolved_sharedtracks_ttks))
+                sv0_resolved_sharedtracks = reco::Vertex(tv);
+
+              sv0 = sv0_resolved_sharedtracks;
+              if (verbose) {
+                printf("sv0'idx: %lu with ntrack after: %u",vtxidx0, sv0.nTracks());
+                printf("sv1'idx: %lu with ntrack after: %u",vtxidx1, sv1.nTracks());
+              }
+            }
+          }
+        }
+      }
+    }
+
+    for (v[0] = vertices->begin(); v[0] != vertices->end(); ++v[0]) {
+      track_set tracks[2];
+      ivtx[0] = v[0] - vertices->begin();
+      tracks[0] = vertex_track_set(*v[0]);
+
+      if (tracks[0].size() < 2) {
+        if (verbose)
+          printf("resolve_shared_jets: vertex-0 #%lu is down to one track, junking it\n", ivtx[0]);
+        v[0] = vertices->erase(v[0]) - 1;
+        ++n_output_aftersharedjets_onetracks;
+        continue;
+      }
+
+      if (tracks[0].size() < v[0]->nTracks()) {
+        std::cout << "duplicate tracks in vertexer" << std::endl;
+        std::vector<reco::TransientTrack> sv_nonduplicate_ttks;
+        for (const reco::TrackRef& itk : vertex_track_set(*v[0])) {
+          if (itk.isNonnull())
+            sv_nonduplicate_ttks.push_back(tt_builder->build(itk));
+        }
+
+        for (const TransientVertex& tv : kv_reco_dropin(sv_nonduplicate_ttks))
+          *v[0] = reco::Vertex(tv);
+      }

```

Comment:
Does this actually get reached when things are behaving as expected? Maybe we should have this throw an error rather than doing the printout and "fixing" the vertex? 

Or maybe fixing it as you've done is fine--what do you think? If this does happen on occasion, how often is it? i.e. will our log files be flooded by the printout, or is it a rare enough occurrence that having this kept track of in the log file is useful?

And to be sure: the `vertex_track_set ` sufficiently removes the duplicate tracks and properly gives us the correct answer, right? 

Last thing: would `tracks[0].size() != v[0]->nTracks()` be sensible here? Perhaps not, since the events with `tracks[0].size()  > v[0]->nTracks() ` would arise differently, and your code wouldn't resolve those. Does this edge case ever happen though?

Reply:
This part of the code was there when I was trying to debug my code in the part of resolve shared jets. I just check the REU ntuple folders with the new vertexing and did not find this printouts at all in the log files. So, I guess we can just remove L1506 to L1530 ? The bug was caused by vertices that propagate to `v[0]->nTracks` so it makes sense that after the bug was fixed, the track set should have the same size as the total tracks.

Reply:
Rather than removing the code, I think having it throw an error as a sanity check when things go wrong is actually quite useful. I'm not sure which L1506-1530 you refer to, since line numbers change with each commit!

Reply:
I changed the printouts in this part to exception. Done.